lib/instrument/transformer/nskeyed.js

```diff
@@ -0,0 +1,310 @@
+import plistlib from 'bplist-parser';
+import bplistCreate from 'bplist-creator';
+import {parse as uuidParse, stringify as uuidStringify} from 'uuid';
+import _ from 'lodash';
+
+const NSKeyedArchiveVersion = 100_000;
+const NULL_UID = new plistlib.UID(0);
+const CYCLE_TOKEN = 'CycleToken';
+const PRIMITIVE_TYPES = ['number', 'string', 'boolean', 'UID', 'Buffer'];
+const NOT_ENCODE_TYPES = ['number', 'boolean'];
+const NSKEYEDARCHIVER = 'NSKeyedArchiver';
+const UNIX2APPLE_TIMESTAMP_SECOND = 978307200.0;
+let decodeArchive;
+let encodeArchive;
+
+function ArchivedObject (object, unarchiver) {
+  this.object = object;
+  this._unarchiver = unarchiver;
+  this.decodeIndex = (index) => this._unarchiver.decodeObject(index);
+  this.decode = (key) => this._unarchiver.decodeKey(this.object, key);
+}
+
+function ArchivingObject (object, archiver) {
+  this._archiveObj = object;
+  this._archiver = archiver;
+  this.encode = (key, val) => {this._archiveObj[key] = this._archiver.encode(val);};
+}
+
+function DictArchive (archive) {
+  const keyUids = archive.decode('NS.keys');
+  const valUids = archive.decode('NS.objects');
+  const d = {};
+  for (let i = 0; i < keyUids.length ; i++) {
+    const key = archive.decodeIndex(keyUids[i]);
+    d[key] = archive.decodeIndex(valUids[i]);
+  }
+  return d;
+}
+
+function ListArchive (archive) {
+  const uids = archive.decode('NS.objects');
+  const _list = [];
+  for (const uid of uids) {
+    _list.push(archive.decodeIndex(uid));
+  }
+  return _list;
+}
+
+function DTTapMessagePlist (archive) {
+  return archive.decode('DTTapMessagePlist');
+}
+
+function NSError (archive) {
+  return {
+    '$class': 'NSError',
+    'domain': archive.decode('NSDomain'),
+    'userinfo': archive.decode('NSUserInfo'),
+    'code': archive.decode('NSCode')};
+}
+
+function NSException (archive) {
+  return {
+    '$class': 'NSException',
+    'reason': archive.decode('NS.reason'),
+    'userinfo': archive.decode('userinfo'),
+    'name': archive.decode('NS.name')};
+}
+
+class NSURL {
+  constructor (base, relative) {
+    this._base = base;
+    this._relative = relative;
+  }
+}
+
+/**
+ * decode and encode Archive of currently known data formats
+ * @type {Readonly}
+ */
+const UNARCHIVE_CLASS_MAP = Object.freeze({
```

Comment:
@mykola-mokhnach could you review the code for me. I learned those websites and refactored here

Reply:
this would probably not work as expected. If you just want to map the particular class then the following needs to be done:

- remove the decodeArchive and encodeArchive global module vars
- `DTSysmonTapMessage: DTTapMessagePlist` should be ok
- Create a base class : `BaseMessagePlist` with two methods `decodeArchive` and `encodeArchive` which implement the default behaviour or throw an exception
- Inherit other classes from it and override the above methods
- The code to extract the necessary method would look like:

```js
    const classUid = rawObj.$class;
    const instance = this.classForUid(classUid)();
    const decodedObj = instance.decodeArchive(new ArchivedObject(rawObj, this));
```

Reply:
Also adding a simple unit test on these method would be helpful to test the actual encoding/decoding behavior

Reply:
In the beginning I planned to do it this way.
But when I study javascript, The document shows that you can simply create a class
```javascript
UNARCHIVE_CLASS_MAP.NSDictionary = function () {
  this.decodeArchive = DictArchive;
};

new UNARCHIVE_CLASS_MAP['NSDictionary']().decodeArchive()
```
I seem to be going in the wrong direction

Reply:
This looks almost the same as python. Let me revise it again

Reply:
yes, I'm also more like a Python person myself. Try to make it simple/easy to read. It is only necessary to use a special javascript dark magic where it is not possible to find other ways.